### ISSUE TITLE
Inspect hidden states for process understanding

### ISSUE THREAD

**Issue by jsadler2** [USGS-R/drb-do-ml#45]
An idea that has come up is to inspect the hidden states to see if they are behaving as we would expect some state or flux in the process would behave. 

The two examples that have come up are:
1. a **biomass**: some representation of a biomass that accumulates in the summer, is lower in the winter, and maybe has a sharp decrease after a scouring event
2. **discharge**: we aren't giving the baseline model discharge as an input. Does the model have a discharge-like hidden state?

We can look into answering these questions with the baseline LSTM model (#40).

**Comment by jsadler2** [comment]
I started looking at this a little bit and it's interesting! A little mysterious, but interesting.

Here are a few plots so far:
## Randomized weights
I first ran the model to get the "DO" predictions and the hidden states with randomized weights to get something to compare them with:

### "DO" preds
![do_rand](https://user-images.githubusercontent.com/6943441/153678772-f34ff458-f696-4b69-8b24-11dc6fa5db57.jpg)

### hidden states
![h_rand](https://user-images.githubusercontent.com/6943441/153678764-ec68d262-cf16-4264-951b-b049ebe757a7.jpg)



## Trained weights
### DO preds
![do](https://user-images.githubusercontent.com/6943441/153678752-a42e97ce-ab1c-47a5-8b93-e6fa1888e5eb.jpg)


### hidden states

![h](https://user-images.githubusercontent.com/6943441/153678735-92e1a833-f78d-4983-8c10-c62a95c80259.jpg)

**Comment by lekoenig** [comment]
Awesome, @jsadler2 - thanks for sharing these! 🚀 

I'm wondering about how/whether I should compare the h time series in the randomized weights plots above versus the trained weights plots below. The weights more or less determine how the inputs are mapped to outputs, right? So in the model run with randomized weights, it seems like the model is learning the relationship between air temperature and DO (e.g. I'm looking at h_2), but that dynamic isn't really maintained in h_2 on the trained weights. It's also interesting that predicted daily mean DO is often greater than the predicted daily max DO for this model run (and that min DO doesn't really have much seasonality) - does that just point to the importance of model training?

For the model run with trained weights, h_3 and h_9 jump out to me as the model appears to be inferring some dynamic that is relatively higher from Oct - May compared to the rest of the year. h_1 looks sort of like a hydrograph (see below), and I can imagine h_0 as some concatenation of the h_3 and h_1. Mysterious but interesting, indeed!

Also interested in your thoughts/interpretation so far, and whether I'm reading these tea leaves appropriately with regard to the randomized versus trained weights. 

![discharge](https://user-images.githubusercontent.com/8785034/153905683-95342b43-521f-4697-964e-b6bf0416609f.png)

**Comment by jsadler2** [comment]
>does that just point to the importance of model training?

Yeah. The randomized weight output is just that - it's totally random. So the seasonal trend we are seeing in h_2 is basically just the trend in the temperature inputs that is randomly coming through. The model hasn't seen any DO data, so it's all just noise. .. and yes. That is especially apparent in the "DO" predictions. They are all over the place and unrealistic (mean > max).

> h_3 and h_9 jump out to me

Those ones stood out to me too. So interesting that there are these two very distinct seasons. That is most apparent in h_3, but I think you could argue that there are two distinct regimes in most all of the states. For example
- h_7 as the negative times and the positive times, 
- h_8 has times with very little variability (winter) and very volatile times (spring - fall). 

> h_1 looks sort of like a hydrograph

I agree. It would be interesting to plot those in the same figure

>whether I'm reading these tea leaves appropriately with regard to the randomized versus trained weights.

That is how I'm reading the tea leaves too :)

One other thing that stood out to me is the sudden drops in h_0. For example, what was it it mid-April 2018 that caused the sudden drop in that state... and there are many similar patterns in h_0, but mid-Apr '18 is the ~biggest magnitude.

**Comment by lekoenig** [comment]
> The model hasn't seen any DO data

Thanks for this explanation, Jeff - that makes a lot more sense. I didn't realize the model hadn't seen any DO data in the randomized version (and that's probably why you had DO in quotes 😃 )

**Comment by galengorski** [comment]
This is a really interesting conversation. It makes me wonder:
1) how consistent are these states from model run to model run at the same site and from site to site
2) is there a way to label or group these in a such a way so that when you run the model again and the new h_7 looks like the old h_0 (for example) we have a way to track that, maybe with entropy or a spectral signature or something...? If I were to group the hidden states of the trained model naively, I would say that h_3, h_4, h_5, and h_9 are dominated by lower frequency signals, while h_2, h_6 and h_7 are dominated by higher frequency, while h_0, h_1, and h_8 are kind of a mix. Maybe at a new site or new model realization, the groupings would be similar even if the numbers are mixed up.
3) what about also tracking the weights of the hidden states? Wouldn't that tell us when these hidden states are more or less important for DO?

**Comment by jsadler2** [comment]
> 1. how consistent are these states from model run to model run at the same site and from site to site

Good question. I'll do a couple runs today to see how they compare across model runs. I can also look across sites.

> 2. is there a way to label or group these

I also am wondering if dynamic time warping would be an good way to measure simliarity.

> 3. what about also tracking the weights of the hidden states? Wouldn't that tell us when these hidden states are more or less important for DO?

Good idea. If I'm understanding this, though, the weights will be just a static 10x3 (or 3x10?) matrix, so I don't think those would tell us anything about the importance in time, but I do think they'd be worth looking at too.

**Comment by lekoenig** [comment]
Jeff has some useful python code in his forked version of the repo for plotting the hidden states. @jsadler2, do you have anything to add here, or any steps you think we should take based on the commits referenced above? Or can we close this issue?
